File: quick_sort_1_index.py
def swap(array, i, j):
    array[i], array[j] = array[j], array[i]


def quickSort(array, left, right):

    if(left >= right):
        return

    last = left
    pivot = (left + right // 2)
    swap(array, left, pivot)
    i = left+1

    while i <= right:
        if array[i] < array[left]:
            last += 1
            swap(array, i, last)
        i += 1

    swap(array, last, left)
    quickSort(array, left, last-1)
    quickSort(array, last+1, right)


array = [4, 3, 5, 32, 1]
quickSort(array, 0, len(array)-1)
for i in array:
    print(i, end="\t")

File: basicAlgrithms/quick_sort_1_index.cpp
#include <iostream>
using namespace std;
void swap(int arr[], int i, int j)
{
    int temp;
    temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

void quickSort(int arr[], int left, int right)
{
    int end = left;
    int pivot = (left + right) / 2;
    swap(arr, left, pivot);
    if (left >= right)
        return;
    for (int i = left + 1; i <= right; i++)
    {
        if (arr[i] < arr[left])
            swap(arr, ++end, i);
    }
    swap(arr, left, end);
    for (int i = 0; i < 4; i++)
    {
        cout << arr[i] << "\t";
    }
    cout << endl;
    quickSort(arr, left, end - 1);
    quickSort(arr, end + 1, right);
}
int main()
{
    int a[] = {4, 45, 3, 22};
    quickSort(a, 0, 3);
}